metadata-ingestion/src/datahub/ingestion/source/s3/source.py

```diff
@@ -938,58 +946,150 @@ def _is_allowed_path(path_spec_: PathSpec, s3_uri: str) -> bool:
                 self.report.report_file_dropped(s3_uri)
             return allowed
 
-        s3_objects = (
-            obj
-            for obj in bucket.objects.filter(Prefix=prefix).page_size(PAGE_SIZE)
-            if _is_allowed_path(
-                path_spec, self.create_s3_path(obj.bucket_name, obj.key)
+        # Process objects in a memory-efficient streaming fashion
+        # Instead of loading all objects into memory, we'll accumulate folder data incrementally
+        folder_data: Dict[
+            str, Dict[str, Any]
+        ] = {}  # dirname -> {"objects": [obj], "total_size": int}
```

Comment:
`total_size` --> `num_objects`, or `objects_length` 
WDYT?

Reply:
`min_time`, `max_time` and `latest_obj` are also in this dictionary, right?

what about defining a dataclass or typeddict for this?